docs/howto/operator/custom-operator.rst

```diff
@@ -0,0 +1,185 @@
+For example, multiple tasks in a DAG can require access to a MySQL database. Instead of
+creating a connection per task, you can retrieve a connection from the hook and utilize it.
+Hook also helps to avoid storing connection auth parameters in a DAG. 
+See :doc:`../connection/index` for how to create and manage connections.
+
+Let's extend our previous example to fetch name from MySQL:
+
+.. code:: python
+
+    class HelloDBOperator(BaseOperator):
+
+            @apply_defaults
+            def __init__(
+                    self,
+                    name: str,
+                    conn_id: str,
+                    database: str,
+                    *args, **kwargs) -> None:
+                super().__init__(*args, **kwargs)
+                self.name = name
+                self.conn_id = conn_id
+                self.database = database
+
+            def execute(self, context):
+                hook = MySqlHook(mysql_conn_id=self.mysql_conn_id,
+                         schema=self.database)
+                sql = "select name from user"
+                result = hook.get_first(sql)
+                message = "Hello {}".format(result['name'])
+                print(message)
+                return message
+
+When the operator invokes the query on the hook object, a new connection gets created if it doesn't exist. 
+The hook retrieves the auth parameters such as username and password from Airflow
+backend and passes the params to the :py:func:`airflow.hooks.base_hook.BaseHook.get_connection`. 
+
+
+User interface
+^^^^^^^^^^^^^^^
+Airflow also allows the developer to control how the operator shows up in the DAG UI.
+Override ``ui_color`` to change the background color of the operator in UI. 
+Override ``ui_fgcolor`` to change the color of the label.
+
+.. code::  python
+
+        class HelloOperator(BaseOperator):
+            ui_color = '#ff0000'
+            ui_fgcolor = '#000000'
+            ....
+
+Templating
+^^^^^^^^^^^
+You can use :ref:`Jinja templates <jinja-templating>` to parameterize your operator.
+Airflow considers the field names present in ``template_fields``  for templating while rendering
+the operator.
+
+.. code:: python
+    
+        class HelloOperator(BaseOperator):
+            
+            template_fields = ['name']
+            
+            @apply_defaults
+            def __init__(
+                    self,
+                    name: str,
+                    *args, **kwargs) -> None:
+                super().__init__(*args, **kwargs)
+                self.name = name
+
+            def execute(self, context):
+                message = "Hello from {}".format(name)
+                print(message)
+                return message
+
+        hello_task = HelloOperator(task_id='task_id_1', dag=dag, name='{{ task_id }}')
+
+In this example, Jinja looks for the ``name`` parameter and substitutes ``{{ task_id }}`` with
+task_id_1.
```

Comment:
```suggestion
``task_id_1``.
```

This substition happens just before the ``execute()`` function is called, so don't do anything in the constructor assuming it will have it's final value.


It might also be worth saying that the template_fields apply to the object attribute names, not the constructor arg names.

I.e. if we do this:

```
            def __init__(
                    self,
                    name: str,
                    *args, **kwargs) -> None:
                super().__init__(*args, **kwargs)
                self._name = name
```

then we'd need `template_fields = ['_name']`